pkg/admission/pod_eviction.go

```diff
@@ -0,0 +1,254 @@
+	otherStatefulSets, err := request.clients.findRelatedStatefulSets(sts, PdbConfig.StsSelector())
+	if err != nil || otherStatefulSets == nil || len(otherStatefulSets.Items) <= 1 {
+		_ = level.Error(request.log).Log(logMsg, "unable to find related stateful sets - continuing with single zone")
+		otherStatefulSets = &appsv1.StatefulSetList{Items: []appsv1.StatefulSet{*sts}}
+	}
+
+	// this is the partition of the pod being evicted - for classic zones the partition will be an empty string
+	partition := PdbConfig.PodPartition(pod)
+	matcher := partitionMatcher{
+		same: func(pd *corev1.Pod) bool {
+			return PdbConfig.PodPartition(pd) == partition
+		},
+	}
+
+	// include the partition we are considering
+	if len(partition) > 0 {
+		request.log.SetSpanAndLogTag(logPartition, partition)
+	}
+
+	// include the number of zones we will be considering
+	request.log.SetSpanAndLogTag(logZones, len(otherStatefulSets.Items))
+
+	singleZone := len(otherStatefulSets.Items) == 1
+
+	for _, otherSts := range otherStatefulSets.Items {
+		// exclude our StatefulSet from this iteration unless we only have found a single zone
+		if !singleZone && otherSts.UID == sts.UID {
+			continue
+		}
+
+		// find the number of not ready pods on this StatefulSet which match the given criteria
+		result, err := request.clients.podsNotRunningAndReady(&otherSts, pod, matcher)
+		if err != nil {
+			// TODO - should we allow or deny this ??
+			_ = level.Error(request.log).Log(logMsg, "unable to determine pod status in related StatefulSet", logStatefulSet, otherSts.Name)
+			continue
+		}
+
+		// Note - if we are in a singleZone mode then we apply the PDB to the single zone and the unavailable pods calculation reflects the state after eviction
+		if result.notReady+result.unknown >= maxUnavailable || (singleZone && result.notReady+result.unknown+1 > maxUnavailable) {
+			_ = level.Info(request.log).Log(logMsg, fmt.Sprintf("%s - pdb exceeded", logDeny), logStatefulSet, otherSts.Name, "notReady", result.notReady, logUnknown, result.unknown, "tested", result.tested, "maxUnavailable", maxUnavailable)
+
+			// Build a nice reason to assist with any debugging
+			msg := ""
+			if result.notReady > 0 {
+				msg += fmt.Sprintf("%d %s not ready", result.notReady, plural(logPod, result.notReady))
+				if result.unknown > 0 {
+					msg += ", "
+				}
+			}
+			if result.unknown > 0 {
+				// 1 pod not ready, 1 pod unknown under ingester-zone-a
+				// 1 pod unknown under ingester-zone-a
+				msg += fmt.Sprintf("%d %s %s %s %s", result.unknown, plural(logPod, result.unknown), logUnknown, logUnder, otherSts.Name)
+			} else if len(partition) > 0 {
+				// 1 pod not ready under ingester-zone-a partition 0
+				msg += fmt.Sprintf(" %s %s %s %s", logUnder, otherSts.Name, logPartition, partition)
+			} else {
+				// 1 pod not ready under ingester-zone-a
+				msg += fmt.Sprintf(" %s %s", logUnder, otherSts.Name)
```

Comment:
I think these format strings would be clearer with the constant parts in the string (eg. `"%d %s unknown under %s"` for the first one)